From 9716b30e586b5b8cec455e41dbd54a38efb06558 Mon Sep 17 00:00:00 2001
From: Marek Ruszczynski <mruszczynski@vewd.com>
Date: Fri, 16 Apr 2021 03:45:52 +0000
Subject: [PATCH] Use mallinfo2 with glibc 2.33 and up

mallinfo2 was added in 2.33 and the previous mallinfo
is now deprecated.

Bug: 1198460
Change-Id: Id0577d1e69a67f0d38618c8cafd08bc30fdabe9d
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/2823312
Reviewed-by: Lei Zhang <thestig@chromium.org>
Commit-Queue: Lei Zhang <thestig@chromium.org>
Cr-Commit-Position: refs/heads/master@{#873183}
---
 base/process/process_metrics_posix.cc | 35 ++++++++++++++++++++++-----
 1 file changed, 29 insertions(+), 6 deletions(-)

diff --git a/base/process/process_metrics_posix.cc b/base/process/process_metrics_posix.cc
index 9d12c427bb30a..587caf48c6651 100644
--- a/base/process/process_metrics_posix.cc
+++ b/base/process/process_metrics_posix.cc
@@ -24,6 +24,10 @@
 #include <malloc.h>
 #endif
 
+#if defined(OS_LINUX) || defined(OS_CHROMEOS) || defined(OS_ANDROID)
+#include <features.h>
+#endif
+
 namespace base {
 
 int64_t TimeValToMicroseconds(const struct timeval& tv) {
@@ -114,18 +118,37 @@ size_t GetPageSize() {
   return pagesize;
 }
 
-size_t ProcessMetrics::GetMallocUsage() {
-#if defined(OS_APPLE)
-  malloc_statistics_t stats = {0};
-  malloc_zone_statistics(nullptr, &stats);
-  return stats.size_in_use;
-#elif defined(OS_LINUX) || defined(OS_CHROMEOS) || defined(OS_ANDROID)
+#if defined(OS_LINUX) || defined(OS_CHROMEOS) || defined(OS_ANDROID)
+namespace {
+
+size_t GetMallocUsageMallinfo() {
+#if defined(__GLIBC__) && defined(__GLIBC_PREREQ)
+#if __GLIBC_PREREQ(2, 33)
+#define MALLINFO2_FOUND_IN_LIBC
+  struct mallinfo2 minfo = mallinfo2();
+#endif
+#endif  // defined(__GLIBC__) && defined(__GLIBC_PREREQ)
+#if !defined(MALLINFO2_FOUND_IN_LIBC)
   struct mallinfo minfo = mallinfo();
+#endif
+#undef MALLINFO2_FOUND_IN_LIBC
 #if BUILDFLAG(USE_TCMALLOC)
   return minfo.uordblks;
 #else
   return minfo.hblkhd + minfo.arena;
 #endif
+}
+
+}  // namespace
+#endif  // defined(OS_LINUX) || defined(OS_CHROMEOS) || defined(OS_ANDROID)
+
+size_t ProcessMetrics::GetMallocUsage() {
+#if defined(OS_APPLE)
+  malloc_statistics_t stats = {0};
+  malloc_zone_statistics(nullptr, &stats);
+  return stats.size_in_use;
+#elif defined(OS_LINUX) || defined(OS_CHROMEOS) || defined(OS_ANDROID)
+  return GetMallocUsageMallinfo();
 #elif defined(OS_FUCHSIA)
   // TODO(fuchsia): Not currently exposed. https://crbug.com/735087.
   return 0;
-- 
2.32.0

